### ISSUE TITLE
Jetbot

### ISSUE THREAD

**Issue by mervo** [AdyRock/com.smartthings#21]
Hey Adrian, really appreciate your custom plugin! Was wondering if I was doing something wrong for my Vacuum Robot. I am unable to trigger it to start cleaning via any of the exposed methods available -> Power/Vol Control; much like how I would toggle it to start from Google Home, of which I am unable to sync my Homey's presence status.

The battery charge status works though!

Thank you!

**Comment by AdyRock** [comment]
Could you enable the log (blue tick) in the Configure App screen. Then try to control the Vacuum Robot and then send me the log.
There is probably another command that I haven't seen before which I will need to add.
I'm not sure how long it will take as my Homey Pro died on Saturday so I am waiting for a new one and as there is a shortage it might take a few weeks.
I will do my best to add it without testing it if it looks straight forward.

**Comment by mervo** [comment]
Wow that's fast Adrian, here's the log:

```
* Setting logEnabled has changed.
* !!!!!! Polling started
* devices/55a3dc44-df1b-3d86-bd4a-a0fae748f37b/components/main/capabilities/audioMute/status
* Get device: devices/55a3dc44-df1b-3d86-bd4a-a0fae748f37b/components/main/capabilities/audioMute/status
Result: {
  "mute": {
    "value": "unmuted",
    "timestamp": "2021-11-18T13:33:35.533Z"
  }
}
* getDeviceValues error: Cannot read property 'value' of undefined
* devices/55a3dc44-df1b-3d86-bd4a-a0fae748f37b/components/main/capabilities/audioVolume/status
* Get device: devices/55a3dc44-df1b-3d86-bd4a-a0fae748f37b/components/main/capabilities/audioVolume/status
Result: {
  "volume": {
    "value": 60,
    "unit": "%",
    "timestamp": "2021-11-19T05:03:42.773Z"
  }
}
* Set Capability: volume_set - Value: 0.6
* devices/55a3dc44-df1b-3d86-bd4a-a0fae748f37b/components/main/capabilities/battery/status
* Get device: devices/55a3dc44-df1b-3d86-bd4a-a0fae748f37b/components/main/capabilities/battery/status
Result: {
  "battery": {
    "value": 100,
    "unit": "%",
    "timestamp": "2021-11-19T09:06:25.961Z"
  }
}
* Set Capability: measure_battery - Value: 100
* devices/55a3dc44-df1b-3d86-bd4a-a0fae748f37b/components/main/capabilities/powerConsumptionReport/status
* Get device: devices/55a3dc44-df1b-3d86-bd4a-a0fae748f37b/components/main/capabilities/powerConsumptionReport/status
Result: {
  "powerConsumption": {
    "value": {
      "energy": 82,
      "deltaEnergy": 0,
      "power": 0,
      "powerEnergy": 0,
      "persistedEnergy": 0,
      "energySaved": 0,
      "start": "2021-11-19T09:05:27Z",
      "end": "2021-11-19T09:20:27Z"
    },
    "timestamp": "2021-11-19T09:20:27.806Z"
  }
}
* Set Capability: measure_power - Value: 0
* Set Capability: meter_power - Value: 0.082
* Set Capability: meter_power.delta - Value: 0
* devices/55a3dc44-df1b-3d86-bd4a-a0fae748f37b/components/main/capabilities/switch/status
* Get device: devices/55a3dc44-df1b-3d86-bd4a-a0fae748f37b/components/main/capabilities/switch/status
Result: {
  "switch": {
    "value": "on",
    "timestamp": "2021-11-18T13:21:17.031Z"
  }
}
* Set Capability: onoff - Value: true
* !!!!!! Polling finished
* Next Interval = 30000
* PostURL url: devices/55a3dc44-df1b-3d86-bd4a-a0fae748f37b/commands
* PostUrl body: {"commands":[{"component":"main","capability":"switch","command":"off","arguments":[]}]}
* HTTPS Error: 409: 
* Vacuum Robot onCapabilityOnoff Error undefined
```

**Comment by AdyRock** [comment]
Hmm, https Error : 409, that in theory means there was a conflict. I'm have no idea what it would be in conflict with though.

In the Homey app, could you first go through the Add Device procedure (no need to actually add anything but get to the screen where it lists the discovered devices) and then send me the contents of More > Apps > SmartThings > Configure App > Detected.
That should show me what capabilities it has so I can see if there is another option to control it.

I will also look into the "getDeviceValues error: Cannot read property 'value' of undefined" to find out what is causing that.

**Comment by mervo** [comment]
cool, I've 2 Smarthings capable devices, my soundbar and vacuum robot which are both picked up.

```
{
  "items": [
    {
      "deviceId": "12627372-6418-f149-37c7-761a19aa2e0b",
      "name": "Living Room Soundbar",
      "label": "Living Room Soundbar",
      "manufacturerName": "Samsung Electronics",
      "presentationId": "VD-NetworkAudio-002S",
      "deviceManufacturerCode": "Samsung Electronics",
      "locationId": "c473da04-4a1c-4521-83a0-fcd7d23e932c",
      "ownerId": "b6acf3d0-f1e7-e8a2-67ee-a39fb66d3eb6",
      "roomId": "e39f52dd-af27-4a97-b3cb-89dfb980dae9",
      "deviceTypeName": "Samsung OCF Network Audio Player",
      "components": [
        {
          "id": "main",
          "label": "main",
          "capabilities": [
            {
              "id": "ocf",
              "version": 1
            },
            {
              "id": "execute",
              "version": 1
            },
            {
              "id": "refresh",
              "version": 1
            },
            {
              "id": "switch",
              "version": 1
            },
            {
              "id": "audioVolume",
              "version": 1
            },
            {
              "id": "audioMute",
              "version": 1
            },
            {
              "id": "audioTrackData",
              "version": 1
            },
            {
              "id": "mediaInputSource",
              "version": 1
            },
            {
              "id": "mediaPlayback",
              "version": 1
            },
            {
              "id": "audioNotification",
              "version": 1
            },
            {
              "id": "samsungvd.soundFrom",
              "version": 1
            },
            {
              "id": "samsungvd.groupInfo",
              "version": 1
            }
          ],
          "categories": [
            {
              "name": "NetworkAudio",
              "categoryType": "manufacturer"
            }
          ]
        }
      ],
      "createTime": "2021-10-10T04:59:09.902Z",
      "profile": {
        "id": "d63d5864-40b7-392b-a5ca-18460598b547"
      },
      "type": "OCF",
      "restrictionTier": 0
    },
    {
      "deviceId": "55a3dc44-df1b-3d86-bd4a-a0fae748f37b",
      "name": "[robot vacuum] Samsung",
      "label": "Vacuum Robot",
      "manufacturerName": "Samsung Electronics",
      "presentationId": "DA-RVC-MAP-01001",
      "deviceManufacturerCode": "Samsung Electronics",
      "locationId": "c473da04-4a1c-4521-83a0-fcd7d23e932c",
      "ownerId": "b6acf3d0-f1e7-e8a2-67ee-a39fb66d3eb6",
      "roomId": "9dfeb2fe-49b0-4279-a9b0-4fe464085b51",
      "deviceTypeName": "Samsung OCF Robot Vacuum",
      "components": [
        {
          "id": "main",
          "label": "main",
          "capabilities": [
            {
              "id": "audioMute",
              "version": 1
            },
            {
              "id": "audioNotification",
              "version": 1
            },
            {
              "id": "audioTrackAddressing",
              "version": 1
            },
            {
              "id": "audioVolume",
              "version": 1
            },
            {
              "id": "battery",
              "version": 1
            },
            {
              "id": "execute",
              "version": 1
            },
            {
              "id": "imageCapture",
              "version": 1
            },
            {
              "id": "mediaPlayback",
              "version": 1
            },
            {
              "id": "mediaTrackControl",
              "version": 1
            },
            {
              "id": "ocf",
              "version": 1
            },
            {
              "id": "powerConsumptionReport",
              "version": 1
            },
            {
              "id": "refresh",

"version": 1
            },
            {
              "id": "robotCleanerCleaningMode",
              "version": 1
            },
            {
              "id": "robotCleanerMovement",
              "version": 1
            },
            {
              "id": "robotCleanerTurboMode",
              "version": 1
            },
            {
              "id": "soundDetection",
              "version": 1
            },
            {
              "id": "switch",
              "version": 1
            },
            {
              "id": "videoCapture",
              "version": 1
            },
            {
              "id": "custom.disabledCapabilities",
              "version": 1
            },
            {
              "id": "custom.disabledComponents",
              "version": 1
            },
            {
              "id": "custom.doNotDisturbMode",
              "version": 1
            },
            {
              "id": "custom.hepaFilter",
              "version": 1
            },
            {
              "id": "samsungce.driverVersion",
              "version": 1
            },
            {
              "id": "samsungce.musicPlaylist",
              "version": 1
            },
            {
              "id": "samsungce.robotCleanerCleaningMode",
              "version": 1
            },
            {
              "id": "samsungce.robotCleanerOperatingState",
              "version": 1
            },
            {
              "id": "samsungce.robotCleanerMapAreaInfo",
              "version": 1
            },
            {
              "id": "samsungce.robotCleanerMapCleaningInfo",
              "version": 1
            },
            {
              "id": "samsungce.robotCleanerPatrol",
              "version": 1
            },
            {
              "id": "samsungce.robotCleanerPetCleaningSchedule",
              "version": 1
            },
            {
              "id": "samsungce.robotCleanerPetMonitor",
              "version": 1
            },
            {
              "id": "samsungce.robotCleanerPetMonitorReport",
              "version": 1
            },
            {
              "id": "samsungce.robotCleanerRelayCleaning",
              "version": 1
            },
            {
              "id": "samsungce.robotCleanerReservation",
              "version": 1
            },
            {
              "id": "samsungce.robotCleanerMotorFilter",
              "version": 1
            },
            {
              "id": "samsungce.soundDetectionSensitivity",
              "version": 1
            }
          ],
          "categories": [
            {
              "name": "RobotCleaner",
              "categoryType": "manufacturer"
            }
          ]
        },
        {
          "id": "station",
          "label": "station",
          "capabilities": [
            {
              "id": "custom.hepaFilter",
              "version": 1
            },
            {
              "id": "samsungce.robotCleanerDustBag",
              "version": 1
            }
          ],
          "categories": [
            {
              "name": "Other",
              "categoryType": "manufacturer"
            }
          ]
        }
      ],
      "createTime": "2021-11-18T13:21:16.123Z",
      "profile": {
        "id": "4f201a24-bdf4-345f-94c5-4ffb83c05734"
      },
      "type": "OCF",
      "restrictionTier": 0
    }
  ],
  "_links": {}
}
```

**Comment by AdyRock** [comment]
OK, It looks like I need to support the robotCleanerCleaningMode and robotCleanerMovement to control the vacuum.
I am also guessing that the cleaner shows options for sound, etc that make no sense?
I presume the cleaner doesn't play music while it is cleaning?
It's strange how Samsung report capabilities for a device that make no sense.

I might make a new device class for the vacuum so that it only shows sensible options and supports the two Components. 'main' and 'station'

To sum up I need to research this one a bit so it might take a bit of time.

**Comment by mervo** [comment]
There's an option to control the cleaner's volume through SmartThings; maybe that's the capability that Samsung is exposing, though the volume control through Homey does not really do anything either.

Please take your time and hope your Homey comes soon! How long did it live for before giving up the ghost?

**Comment by AdyRock** [comment]
This was my second Homey as I upgraded it to the Pro (as it used to be called) and I think that was about 4 years ago.
My friend is still using my original one so that must be about 6 to 7 years old. 

To be honest the memory on my Homey gets a tremendous bashing through constantly uploading development code so I probably reduced its life span.

**Comment by AdyRock** [comment]
I have add support for the cleaning mode, cleaning movement and turbo mode and publish an update in the HCS.
I have no way to test it so no guarantees but give it a go and let me know.
You will need to delete the old vacuum from homey and add it again to get the new capabilities.

**Comment by mervo** [comment]
Brilliant! I am able to do what I set out to do -> Home the cleaner and initiate cleaning via flows. Hats off to you!
